keybase/config.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2016 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package keybase
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/keybase/go-logging"
+	"github.com/keybase/go-updater"
+	"github.com/keybase/go-updater/util"
+)
+
+type config struct {
+	// appName is the name of the app, e.g. "Keybase"
+	appName string
+	// pathToKeybase is the location of the keybase executable
+	pathToKeybase string
+	// log is the logging location
+	log logging.Logger
+	// store is the config values
+	store store
+}
+
+// store is the config values
+type store struct {
+	InstallID string `json:"installId"`
+	Auto      bool   `json:"auto"`
+	AutoSet   bool   `json:"autoSet"`
+}
+
+// newConfig loads a config, which is valid even if it has an error
+func newConfig(appName string, pathToKeybase string, log logging.Logger) (config, error) {
+	cfg := newDefaultConfig(appName, pathToKeybase, log)
+	err := cfg.load()
+	return cfg, err
+}
+
+func newDefaultConfig(appName string, pathToKeybase string, log logging.Logger) config {
+	return config{
+		appName:       appName,
+		pathToKeybase: pathToKeybase,
+		log:           log,
+	}
+}
+
+// load the config
+func (c *config) load() error {
+	path, err := c.path()
+	if err != nil {
+		return nil
+	}
+	return c.loadFromPath(path)
+}
+
+func (c *config) loadFromPath(path string) error {
+	if _, serr := os.Stat(path); os.IsNotExist(serr) {
+		c.log.Warningf("Unable to load config, %s doesn't exist", path)
+		return serr
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	if file == nil {
+		return fmt.Errorf("No file")
+	}
+	defer func() { _ = file.Close() }()
+
+	decoder := json.NewDecoder(file)
+	if err := decoder.Decode(&c.store); err != nil {
```

Comment:
Yeah it is streaming so it will mutate the store as it parses...

This is a good point, perhaps we want to to revert to the default state in case of any parse error.

I made this change and added a test for this.

We can change our mind though if we want config to mutate while parsing and we decide we want all valid state we can get?